staging/src/k8s.io/client-go/transport/websocket/client.go

```diff
@@ -160,65 +120,109 @@ func (rt *RoundTripper) RoundTrip(request *http.Request) (retResp *http.Response
-// as a pointer to the websocket RoundTripper. The websocket RoundTripper contains the
-// websocket connection after RoundTrip() on the wrapper. Returns an error if there is
-// a problem creating the round trippers.
-func RoundTripperFor(config *restclient.Config) (http.RoundTripper, ConnectionHolder, error) {
+// Client knows how to establish a connection to a remote WebSocket endpoint and make it available for use.
+type Client struct {
+	rt        *wsRoundtripper
+	rtUpgrade http.RoundTripper
+}
+
+// Connection returns the stored websocket connection.
+func (c *Client) Connection() *gwebsocket.Conn {
+	return c.rt.conn
+}
+
+// DataBufferSize returns the size of buffers for the
+// websocket connection.
+func (c *Client) DataBufferSize() int {
+	return dataBufferSize
+}
+
+// TLSClientConfig implements pkg/util/net.TLSClientConfigHolder.
+func (c *Client) TLSClientConfig() *tls.Config {
+	return c.rt.dialer.TLSClientConfig
+}
+
+// NewClient from the restclient.Config.
+// The websocket Client contains the websocket connection after the connection is established.
+// Returns an error if there is a problem creating the round trippers.
+func NewClient(config *restclient.Config) (*Client, error) {
 	transportCfg, err := config.TransportConfig()
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 	tlsConfig, err := transport.TLSConfigFor(transportCfg)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 	proxy := config.Proxy
 	if proxy == nil {
 		proxy = utilnet.NewProxierWithNoProxyCIDR(http.ProxyFromEnvironment)
 	}
 
-	upgradeRoundTripper := &RoundTripper{
-		TLSConfig: tlsConfig,
-		Proxier:   proxy,
+	dialer := gwebsocket.Dialer{
+		Proxy:           proxy,
+		TLSClientConfig: tlsConfig,
+		ReadBufferSize:  dataBufferSize + 1024, // add space for the protocol byte indicating which channel the data is for
+		WriteBufferSize: dataBufferSize + 1024, // add space for the protocol byte indicating which channel the data is for
+	}
+
+	if config.Dial != nil {
+		dialer.NetDialContext = config.Dial
 	}
-	wrapper, err := transport.HTTPWrappersForConfig(transportCfg, upgradeRoundTripper)
+
+	rt := &wsRoundtripper{
+		dialer: &dialer,
+	}
+
+	wrapper, err := transport.HTTPWrappersForConfig(transportCfg, rt)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
+	}
+
+	client := &Client{
+		rt:        rt,
+		rtUpgrade: wrapper,
 	}
-	return wrapper, upgradeRoundTripper, nil
+
+	return client, nil
 }
 
-// Negotiate opens a connection to a remote server and attempts to negotiate
+// Connect opens a connection to a remote server and attempts to negotiate
 // a WebSocket connection. Upon success, it returns the negotiated connection.
 // The round tripper rt must use the WebSocket round tripper wsRt - see RoundTripperFor.
-func Negotiate(rt http.RoundTripper, connectionInfo ConnectionHolder, req *http.Request, protocols ...string) (*gwebsocket.Conn, error) {
```

Comment:
This function has a dependency on rt http.RoundTripper, connectionInfo ConnectionHolder to be generated on certain way, so it is better to make them part of the object